### PR TITLE
feat(treeview): initial implementation of tree view

### PR DESCRIPTION
**Pull request by joginder07**
#### Description

* Initial implementation of the tree view component
* Added documentation
* Will add tests once implementation is reviewed

#### Scope

- [ ] Patch: Bug Fix
- [x] Minor: New Feature
- [ ] Major: Breaking Change

### CODE REVIEW

File: documentation-site/examples/tree-view/controlled.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    ],
+  },
+];
+
+export default function TreeViewControlled() {
+  const [data, setData] = React.useState(initialData);
+
+  const onToggle = (node: TreeNodeT) => {
+    setData(prevData => {
+      node.isExpanded = !node.isExpanded;
```
Comment (chasestarr) [inline documentation-site/examples/tree-view/controlled.js]: This feels like it breaks the intuition of React's state update model. The reason why it works is that the node in-scope references an element in the array provided to the component. The approach relies too much on the implementation details of the `TreeView` component and assumes that the reference is always stable. This code will break if the component needs to perform transformations on the array. Can you please refactor these examples to a more immutable approach?

File: src/tree-view/types.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+import * as React from 'react';
+import type {OverrideT} from '../helpers/overrides.js';
+
+export type StylePropsT = {
+  $isChildNode?: boolean,
+};
+
+export type TreeViewOverridesT = {
+  Root?: OverrideT<StylePropsT>,
+  TreeItemList?: OverrideT<StylePropsT>,
```
[the remark this answers is not in the document]
Comment (chasestarr) [reply src/tree-view/types.js]: My suggestion is to type the overrides like below. The `StylePropsT` type should not be defined on components that have no intention of using `$isChildNode` or `$isLeafNode`.

```js
export type TreeViewOverridesT = {
  Root?: OverrideT<{}>,
  TreeItemList?: OverrideT<{$isChildNode?: boolean}>,
  TreeItem?: OverrideT<{$isLeafNode?: boolean}>,
  TreeItemContent?: OverrideT<{}>,
  IconContainer?: OverrideT<{}>,
  ExpandIcon?: OverrideT<{}>,
  CollapseIcon?: OverrideT<{}>,
};
```

File: src/tree-view/__tests__/tree-view.scenario.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        ],
+      },
+    ],
+  },
+];
+
+export const component = () => (
+  <StatefulTreeView
+    data={initialData}
+    overrides={{
```
[the remark this answers is not in the document]
Comment (joginder07) [reply src/tree-view/__tests__/tree-view.scenario.js]: Actually, it's not required. I will remove this.

**Comment by joginder07** [comment]
> Thanks, looking great! I appreciate you adding tests. Can you please review my comment related to override types above? Other than that, and a small nit, LGTM

@chasestarr  Thanks for the review. Updated.

File: src/tree-view/styled-components.js [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+*/
+
+// @flow
+
+import {styled} from '../styles/index.js';
+
+import type {StylePropsT} from './types.js';
+
+// $FlowFixMe https://github.com/facebook/flow/issues/7745
+export const StyledTreeItemList = styled<StylePropsT>(
```
Comment (chasestarr) [inline src/tree-view/styled-components.js]:
```suggestion
export const StyledTreeItemList = styled<{$isChildNode?: boolean}>(
```

File: src/tree-view/styled-components.js [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+      paddingBottom: 0,
+      paddingLeft: 0,
+      paddingRight: 0,
+      position: 'relative',
+    };
+  },
+);
+
+// $FlowFixMe https://github.com/facebook/flow/issues/7745
+export const StyledTreeItem = styled<StylePropsT>(
```
Comment (chasestarr) [inline src/tree-view/styled-components.js]:
```suggestion
export const StyledTreeItem = styled<{$isLeafNode?: boolean}>(
```

File: src/tree-view/styled-components.js [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+      paddingBottom: 0,
+      paddingLeft: 0,
+      paddingRight: 0,
+      position: 'relative',
+    };
+  },
+);
+
+// $FlowFixMe https://github.com/facebook/flow/issues/7745
+export const StyledItemContent = styled<StylePropsT>('div', ({$theme}) => {
```
Comment (chasestarr) [inline src/tree-view/styled-components.js]:
```suggestion
export const StyledItemContent = styled<{}>('div', ({$theme}) => {
```

File: src/tree-view/styled-components.js [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    marginLeft: 0,
+    marginRight: 0,
+    ':hover': {
+      backgroundColor: $theme.colors.mono300,
+    },
+  };
+});
+
+// $FlowFixMe https://github.com/facebook/flow/issues/7745
+export const StyledIconContainer = styled<StylePropsT>('div', ({$theme}) => {
```
Comment (chasestarr) [inline src/tree-view/styled-components.js]:
```suggestion
export const StyledIconContainer = styled<{}>('div', ({$theme}) => {
```

File: src/tree-view/styled-components.js [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+This source code is licensed under the MIT license found in the
+LICENSE file in the root directory of this source tree.
+*/
+
+// @flow
+
+import {styled} from '../styles/index.js';
+
+import type {StylePropsT} from './types.js';
```
Comment (chasestarr) [inline src/tree-view/styled-components.js]: Please remove this type

File: src/tree-view/types.js [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+LICENSE file in the root directory of this source tree.
+*/
+// @flow
+import * as React from 'react';
+import type {OverrideT} from '../helpers/overrides.js';
+
+export type StylePropsT = {
+  $isChildNode?: boolean,
+  $isLeafNode?: boolean,
+};
```
Comment (chasestarr) [inline src/tree-view/types.js]: Please remove this type

**Comment by chasestarr** [comment]
Visual regression tests failed since there is now a new `.scenario.js` file. Can you please follow [these steps](https://github.com/uber/baseweb/tree/master/vrt#docker) to update the VRT snapshots?

**Comment by chasestarr** [comment]
Thank you @joginder07 !

**Pull request merged**